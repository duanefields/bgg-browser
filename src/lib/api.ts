/** Root URL of the BGG Proxy Server */
const BGG_PROXY: string = import.meta.env.VITE_BGG_PROXY

type CollectionResponse = {
  items: {
    item: {
      _objectid: number
      name: {
        _sortindex: string
        __text: string
      }
      yearpublished: string
      image: string
      thumbnail: string
      stats: {
        _minplayers: string
        _maxplayers: string
        _minplaytime: string
        _maxplaytime: string
        _playingtime: string
        _numowned: string
        rating: {
          usersrated: {
            _value: string
          }
          average: {
            _value: string
          }
          bayesaverage: {
            _value: string
          }
          stddev: {
            _value: string
          }
          median: {
            _value: string
          }
          ranks: {
            rank: {
              _type: string
              _id: string
              _name: string
              _friendlyname: string
              _value: string
              _bayesaverage: string
            }[]
            _value: string
          }
        }
      }
      status: {
        _own: string
        _prevowned: string
        _fortrade: string
        _want: string
        _wanttoplay: string
        _wanttobuy: string
        _wishlist: string
        _preordered: string
        _lastmodified: string
      }
      numplays: string
      _objecttype: string
      _subtype: string
      _collid: string
    }[]
  }
  _totalitems: string
  _termsofuse: string
  _pubdate: string
}

type Game = {
  objectId: number
  name: string
  thumbnail: string
  minPlayers: number
  maxPlayers: number
  minPlaytime: number
  maxPlaytime: number
  playtime: number
  yearPublished: number
  averageRating: number
  // rank: number
  numPlays: number
}

const getCollection = async (username: string): Promise<Game[]> => {
  const response = await fetch(
    `${BGG_PROXY}/collection/?username=${username}&stats=1&own=1&version=1&excludesubtype=boardgameexpansion`,
  )

  // Check for errors, the server responds with a 200 OK even if there is an error
  const json = await response.json()
  if (json.errors) {
    throw new Error(json.errors.error.message)
  }

  const data = json as CollectionResponse
  return data.items.item.map((item) => ({
    objectId: Number(item._objectid),
    name: item.name.__text,
    thumbnail: item.thumbnail,
    minPlayers: Number(item.stats._minplayers || 0),
    maxPlayers: Number(item.stats._maxplayers || 0),
    minPlaytime: Number(item.stats._minplaytime || 0),
    maxPlaytime: Number(item.stats._maxplaytime || 0),
    playtime: Number(item.stats._playingtime || 0),
    yearPublished: Number(item.yearpublished),
    averageRating: Number(Number(item.stats.rating.average._value || 0).toFixed(1)),
    // rank: item.stats.rank,
    numPlays: Number(item.numplays),
  }))
}

export { getCollection, BGG_PROXY }
