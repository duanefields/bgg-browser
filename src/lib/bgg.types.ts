export interface RankResponse {
  _type: string
  _id: string
  _name: string
  _friendlyname: string
  _value: string
  _bayesaverage: string
}

export interface ItemResponse {
  _objectid: number
  name: {
    _sortindex: string
    __text: string
  }
  yearpublished: string
  image: string
  thumbnail: string
  stats: {
    _minplayers: string
    _maxplayers?: string
    _minplaytime?: string
    _maxplaytime?: string
    _playingtime?: string
    _numowned: string
    rating: {
      usersrated: {
        _value: string
      }
      average: {
        _value: string
      }
      bayesaverage: {
        _value: string
      }
      stddev: {
        _value: string
      }
      median: {
        _value: string
      }
      ranks: {
        rank: RankResponse[] | { _value: string }
        _value: string
      }
      _value: string
    }
  }
  status: {
    _own: string
    _prevowned: string
    _fortrade: string
    _want: string
    _wanttoplay: string
    _wanttobuy: string
    _wishlist: string
    _preordered: string
    _lastmodified: string
  }
  numplays: string
  _objecttype: string
  _subtype: string
  _collid: string
}

export interface CollectionResponse {
  items: {
    item?: ItemResponse[]
  }
  _totalitems: string
  _termsofuse: string
  _pubdate: string
}

export interface UserResponse {
  user: {
    firstname: { _value: string }
    lastname: { _value: string }
    avatarlink: { _value: string }
    yearregistered: { _value: string }
    lastlogin: { _value: string }
    stateorprovince: { _value: string }
    country: { _value: string }
    webaddress: { _value: string }
    xboxaccount: { _value: string }
    wiiaccount: { _value: string }
    psnaccount: { _value: string }
    battlenetaccount: { _value: string }
    steamaccount: { _value: string }
    traderating: { _value: string }
    _id: string
    _name: string
    _termsofuse: string
  }
}

export interface InvalidUsernameResponse {
  errors: {
    error: {
      message: string
    }
  }
}

export interface CollectionProcessingResponse {
  message: string
}
