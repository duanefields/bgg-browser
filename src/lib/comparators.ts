import { Game } from "../shared.types"

/** sort by name ascending, ignoring leading articles a, an, and the */
export const titleComparator = (a: Game, b: Game) => {
  const aTitle = a.name.replace(/^(a|an|the) /i, "")
  const bTitle = b.name.replace(/^(a|an|the) /i, "")
  return aTitle.localeCompare(bTitle)
}

/** sort by averageRating descending, then by title */
export const ratingComparator = (a: Game, b: Game) => {
  return b.averageRating - a.averageRating || titleComparator(a, b)
}

/** sort by myRating descending, then by title */
export const myRatingComparator = (a: Game, b: Game) => {
  return (b.myRating || 0) - (a.myRating || 0) || titleComparator(a, b)
}

/** sort by numPlays descending, then by title */
export const playsComparator = (a: Game, b: Game) => {
  return b.numPlays - a.numPlays || titleComparator(a, b)
}

/** sort by rank ascending, then by title (though rank should be unique) */
export const rankComparator = (a: Game, b: Game) => {
  return (
    (a.rank || Number.MAX_SAFE_INTEGER) - (b.rank || Number.MAX_SAFE_INTEGER) ||
    titleComparator(a, b)
  )
}
