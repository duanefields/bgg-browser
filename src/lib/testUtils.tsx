import { QueryClient, QueryClientProvider } from "@tanstack/react-query"
import { createRootRoute, createRoute, createRouter, RouterProvider } from "@tanstack/react-router"
import { render } from "@testing-library/react"

/** Renders a collection of children with a Route and QueryClient Provider */
export const renderWithProviders = (children: React.ReactNode) => {
  // wrap the component in a fresh QueryClientProvider to isolate the tests
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  // create a router with a single route that renders the children in a QueryClientProvider
  const rootRoute = createRootRoute()
  const indexRoute = createRoute({
    getParentRoute: () => rootRoute,
    path: "/",
    component: () => <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>,
  })
  const router = createRouter({
    routeTree: rootRoute.addChildren([indexRoute]),
  })

  return render(<RouterProvider router={router} />)
}

/** Renders a collection of children with just a QueryClient Provider */
export const renderWithQueryProvider = (children: React.ReactNode) => {
  // wrap the component in a fresh QueryClientProvider to isolate the tests
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  return render(<QueryClientProvider client={queryClient}>{children}</QueryClientProvider>)
}
