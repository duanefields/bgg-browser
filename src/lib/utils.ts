import { Game } from "../shared.types"

/** Randomly shuffle an array of games */
export const shuffle = (games: Game[]) => {
  return games
    .map((value) => ({ value, sort: Math.random() }))
    .sort((a, b) => a.sort - b.sort)
    .map(({ value }) => value)
}
