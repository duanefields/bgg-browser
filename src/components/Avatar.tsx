import { useQuery } from "@tanstack/react-query"
import { getUser } from "../lib/api"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { faUser } from "@fortawesome/free-solid-svg-icons"
import classes from "./Avatar.module.css"

interface AvatarProps {
  /** The username of the avatar you want to load */
  username: string
}

const Avatar = ({ username }: AvatarProps) => {
  const query = useQuery({
    queryKey: ["user", username],
    queryFn: () => getUser(username),
    retry: (failureCount, error) => {
      console.debug(`Error #${failureCount} fetching user:`, error.message)
      return error.message !== "Invalid username specified"
    },
  })

  // not everyone has there name set, so use the username as a fallback
  let title = username
  if (query.isSuccess && query.data?.firstName && query.data?.lastName) {
    title = `${query.data.firstName} ${query.data.lastName}`
  }

  // if the user doesn't have an avatar or there is an error, render a default icon
  if ((query.isSuccess && query.data?.avatar === null) || query.isError || query.isLoading) {
    return (
      <FontAwesomeIcon
        icon={faUser}
        width={64}
        height={64}
        size="4x"
        title={title}
        titleId="1000" // this is to ensure snapshots are consistent
        className={classes.avatar}
      />
    )
  }

  return (
    <div>
      {query.isSuccess && query.data?.avatar && (
        <img
          className={classes.avatar}
          src={query.data?.avatar}
          alt={username}
          title={title}
          width={64}
          height={64}
        />
      )}
    </div>
  )
}

export default Avatar
