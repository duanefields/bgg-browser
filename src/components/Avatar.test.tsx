import { screen } from "@testing-library/react"
import { http, HttpResponse } from "msw"
import { describe, expect, it } from "vitest"
import { BGG_PROXY } from "../lib/api"
import { renderWithQueryProvider } from "../lib/testUtils"
import server from "../mockServer"

import invalidUsername from "../test/users/invalidUsername.json"

import Avatar from "./Avatar"

describe("Avatar", () => {
  it("Should match snapshot for image", async () => {
    const view = renderWithQueryProvider(<Avatar username="dkf2112" size={64} />)
    await screen.findByTitle("Duane Fields")
    expect(view).toMatchSnapshot()
  })

  it("Should match snapshot for initials", async () => {
    const view = renderWithQueryProvider(<Avatar username="AlexvW" size={64} />)
    await screen.findByTitle("Alex van Wijngaarden")
    expect(view).toMatchSnapshot()
  })

  it("Should render the user's avatar image if they have one", async () => {
    renderWithQueryProvider(<Avatar username="pandyandy" />)
    const avatarElement = await screen.findByAltText("pandyandy")
    expect(avatarElement.tagName).toBe("IMG")
    expect(avatarElement).toHaveAttribute(
      "src",
      "https://cf.geekdo-static.com/avatars/avatar_id180271.jpg",
    )
    expect(avatarElement).toBeVisible()
  })

  it("Should render initials if they don't have an avatar", async () => {
    renderWithQueryProvider(<Avatar username="AlexvW" />)
    const avatarElement = await screen.findByTitle("Alex van Wijngaarden")
    expect(avatarElement).toHaveTextContent("AV")
    expect(avatarElement).toBeVisible()
  })

  it("Should render a skeleton while loading", () => {
    renderWithQueryProvider(<Avatar username="AlexvW" />)
    const avatarElement = screen.getByTestId("skeleton")
    expect(avatarElement).toBeVisible()
  })

  it("Should render username initial for an invalid username", async () => {
    renderWithQueryProvider(<Avatar username="invalidUsername" />)
    const avatarElement = await screen.findByTitle("invalidUsername")
    expect(avatarElement).toHaveTextContent("I")
    expect(avatarElement).toBeVisible()
  })

  it("Should not retry the user fetch on invalid user error", () => {
    // prepend a handler to the server that will return an invalid user error
    server.use(
      http.get(
        `${BGG_PROXY}/user`,
        () => {
          return HttpResponse.json(invalidUsername)
        },
        { once: true },
      ),
    )

    renderWithQueryProvider(<Avatar username="pandyandy" />)
    // use `queryBy` to avoid throwing an error with `getBy`
    expect(screen.queryByAltText("pandyandy")).not.toBeInTheDocument()
  })

  it("Should render the user's name as the title on the initials if they have it set", async () => {
    renderWithQueryProvider(<Avatar username="AlexvW" />)
    const avatarElement = await screen.findByTitle("Alex van Wijngaarden")
    expect(avatarElement).toBeVisible()
  })

  it("Should render the user's name as the title on the image if they have it set", async () => {
    renderWithQueryProvider(<Avatar username="dkf2112" />)
    const avatarElement = await screen.findByTitle("Duane Fields")
    expect(avatarElement).toBeVisible()
  })

  it("Should render the username as the title on the icon if the user has no name set", async () => {
    renderWithQueryProvider(<Avatar username="pandyandy" />)
    const avatarElement = await screen.findByTitle("pandyandy")
    expect(avatarElement).toBeVisible()
  })

  it("Should render the username as the alt text on the image", async () => {
    renderWithQueryProvider(<Avatar username="dkf2112" />)
    const avatarElement = await screen.findByAltText("dkf2112")
    expect(avatarElement).toBeVisible()
  })
})
