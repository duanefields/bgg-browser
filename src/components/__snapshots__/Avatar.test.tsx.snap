// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar > Should match snapshot for image 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiAvatar-root MuiAvatar-circular _avatar_65174a css-7pbt6e-MuiAvatar-root"
        title="Duane Fields"
      >
        <img
          alt="dkf2112"
          class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
          src="https://cf.geekdo-static.com/avatars/avatar_id95740.jpg"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiAvatar-root MuiAvatar-circular _avatar_65174a css-7pbt6e-MuiAvatar-root"
      title="Duane Fields"
    >
      <img
        alt="dkf2112"
        class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
        src="https://cf.geekdo-static.com/avatars/avatar_id95740.jpg"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Avatar > Should match snapshot for initials 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault _avatar_65174a css-wheijw-MuiAvatar-root"
        title="Alex van Wijngaarden"
      >
        AV
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault _avatar_65174a css-wheijw-MuiAvatar-root"
      title="Alex van Wijngaarden"
    >
      AV
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
